### PR TITLE
Revamp List.ascii_printable?/2

### PR DESCRIPTION
**Pull request by eksperimental**
- Simplify logic
- Add specs
- Validate specs with guards, raising on non-lists
- Add tests
- Correct documentation about improper lists
- Rename argument `counter` to `limit`

### CODE REVIEW

**Comment by eksperimental** [comment]
Please let me know if the tests are missing an edge case

**Comment by josevalim** [comment]
Can we please keep the implementation the same? The compiler can optimize
multiple clauses like it was before better. The current version will do
linear lookups.
-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

File: lib/elixir/lib/list.ex [hunk 1]
```diff
@@ -467,62 +467,44 @@ defmodule List do
...
+        when limit: :infinity | non_neg_integer
+  def ascii_printable?(list, limit \\ :infinity)
+      when is_list(list) and (limit == :infinity or (is_integer(limit) and limit >= 0)) do
+    ascii_printable_with_limit?(list, limit)
   end
 
-  def ascii_printable?([?\b | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
+  @compile {:inline, ascii_printable_with_limit?: 2}
```
Comment (michalmuskala) [inline lib/elixir/lib/list.ex]: inlining recursive functions usually is not very beneficial

Comment (michalmuskala) [reply lib/elixir/lib/list.ex]: in general, it's highly specific from case to case, but overall, it generally doesn't provide huge benefits. You'd need to measure it

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: thank you

File: lib/elixir/lib/list.ex [hunk 2]
```diff
@@ -467,62 +467,43 @@ defmodule List do
...
-  end
+  defp ascii_printable_with_limit?(_list, 0),
+    do: true
 
-  def ascii_printable?([?\e | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
+  for control_char <- [?\a, ?\b, ?\e, ?\f, ?\n, ?\r, ?\t, ?\v] do
+    defp ascii_printable_with_limit?([unquote(control_char) | rest], limit),
+      do: ascii_printable_with_limit?(rest, decrement(limit))
   end
```
Comment (eksperimental) [inline lib/elixir/lib/list.ex]: It is not compiling (it was because I didn't run `make clean`).
How can I call a list comprehension in Erlang, from Elixir? or any why to bypass this limitation?

Comment (josevalim) [reply lib/elixir/lib/list.ex]: This is a bootstrap issue. We can't use comprehensions on the List module because we likely need the List module for comprehensions to work (or something else that depends on the list). We can't be "smart" here.

But as I said in another comment, let's please keep the implementation as is in this commit and only change the minimum necessary to add guards (i.e. only rename the def/calls, no change to variables or syntax).

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: done

File: lib/elixir/lib/list.ex [hunk 3]
```diff
@@ -467,58 +467,66 @@ defmodule List do
       iex> List.ascii_printable?('abc' ++ [0], 2)
       true
 
-  Improper lists are not printable, even if made only of ascii characters:
+  Last element in an improper list renders them not printable, even if made only of ASCII characters:
```
Comment (josevalim) [inline lib/elixir/lib/list.ex]: I am having trouble understanding the rewritten sentence. I will let others give their feedback too. :)

Comment (fertapric) [reply lib/elixir/lib/list.ex]: Same here. I think the previous version was easier to understand. I would only change `ascii` -> `ASCII`

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: Is there a specific term to refer to the last element in an improper list?
What I am trying to say is that if we reach this one element within `limit`, this the the one that renders it non-printable

Comment (eksperimental) [reply lib/elixir/lib/list.ex]: Note that I am correct the previous sentence, since it is not correct.
Improper lists are printable provided limit doesn't reach their last element

Comment (josevalim) [reply lib/elixir/lib/list.ex]: It is the same as the other discussion. :) They are not printable. If you pass one, then it may work by accident, but it is not a guarantee our API provides.

File: lib/elixir/test/elixir/list_test.exs [hunk 4]
```diff
@@ -257,4 +257,39 @@ defmodule ListTest do
...
+      assert List.ascii_printable?('abc' ++ [0], 2)
+    end
+
+    test "improper lists" do
+      refute List.ascii_printable?('abc' ++ ?d)
+      assert List.ascii_printable?('abc' ++ ?d, 3)
+      assert List.ascii_printable?([1, 2 | 3], 0)
+    end
+
+    test "function clause errors" do
```
Comment (josevalim) [inline lib/elixir/test/elixir/list_test.exs]: We generally don't test function clause errors.

Comment (eksperimental) [reply lib/elixir/test/elixir/list_test.exs]: should i spread those tests around the other test blocks? or remove the tests

**Comment by josevalim** [comment]
Regardless of being the same case or not, we don’t document things that
work by accident. The spec should be clear, it only works with lists, we
are not required to check it, which is why we also don’t check for function
clause errors.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by eksperimental** [comment]
I was reviewing the List module, and a lot of those functions break when passed an improper list. I will revert those changes. thank you

**Comment by josevalim** [comment]
@michalmuskala i don't see any merge conflicts in the UI? :D

**Comment by josevalim** [comment]
Oh, doh, I resolved them apparently.

**Comment by michalmuskala** [comment]
Thank you! :heart:
